### PR TITLE
p2p: Move peer-discovery out of PeerPool

### PR DESCRIPTION
**Pull request by gsalgado**
We now have a separate service (DiscoveryService) that performs
discovery, when needed, and adds the peers to its PeerPool

Motivated by the discussion in #886, but also because discovery
shouldn't really be the responsibility of PeerPool

### CODE REVIEW

File: p2p/discovery.py [hunk 1]
```diff
@@ -93,7 +100,13 @@ def __init__(self,
     async def lookup_random(self, cancel_token: CancelToken) -> List[kademlia.Node]:
         return await self.kademlia.lookup_random(self.cancel_token.chain(cancel_token))
 
-    def get_random_nodes(self, count: int) -> Generator[kademlia.Node, None, None]:
+    def get_random_bootnode(self) -> Generator[kademlia.Node, None, None]:
```
Comment (cburgdorf) [inline p2p/discovery.py]: `def get_random_bootnode(self) -> Iterable[kademlia.Node]` seems like a better fit for the signature

File: p2p/discovery.py [hunk 2]
```diff
@@ -93,7 +100,13 @@ def __init__(self,
...
+        if self.bootstrap_nodes:
+            yield random.choice(self.bootstrap_nodes)
+        else:
+            self.logger.warning('No bootnodes available')
+
+    def get_nodes_to_connect(self, count: int) -> Generator[kademlia.Node, None, None]:
```
Comment (cburgdorf) [inline p2p/discovery.py]: Same here, can we make that `Iterable[kademlia.Node]`

File: p2p/discovery.py [hunk 3]
```diff
@@ -234,6 +247,131 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
...
+        if preferred_nodes is not None:
+            self.preferred_nodes = preferred_nodes
+        else:
+            self.logger.debug('Operating with no preferred nodes')
+            self.preferred_nodes = tuple()
+
+        self._preferred_node_tracker = collections.defaultdict(lambda: 0)
+
+    @to_tuple
+    def _get_eligible_preferred_nodes(self) -> Generator[kademlia.Node, None, None]:
```
Comment (cburgdorf) [inline p2p/discovery.py]: Same here, `Iterable[kademlia.Node]`

File: p2p/discovery.py [hunk 4]
```diff
@@ -234,6 +247,131 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
...
+        """
+        Return a random node from the preferred list.
+        """
+        eligible_nodes = self._get_eligible_preferred_nodes()
+        if not eligible_nodes:
+            raise NoEligibleNodes("No eligible preferred nodes available")
+        node = random.choice(eligible_nodes)
+        return node
+
+    def get_random_bootnode(self) -> Generator[kademlia.Node, None, None]:
```
Comment (cburgdorf) [inline p2p/discovery.py]: and here `Iterable[kademlia.Node]`

File: p2p/discovery.py [hunk 5]
```diff
@@ -234,6 +247,131 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
...
+        Return a single node to bootstrap, preferring nodes from the preferred list.
+        """
+        try:
+            node = self._get_random_preferred_node()
+            self._preferred_node_tracker[node] = time.time()
+            yield node
+        except NoEligibleNodes:
+            yield from super().get_random_bootnode()
+
+    def get_nodes_to_connect(self, count: int) -> Generator[kademlia.Node, None, None]:
```
Comment (cburgdorf) [inline p2p/discovery.py]: and here `Iterable[kademlia.Node]`

File: p2p/discovery.py [hunk 6]
```diff
@@ -234,9 +247,134 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
...
+                 privkey: datatypes.PrivateKey,
+                 address: kademlia.Address,
+                 bootstrap_nodes: Tuple[kademlia.Node, ...],
+                 preferred_nodes: Sequence[kademlia.Node] = None) -> None:
+        super().__init__(privkey, address, bootstrap_nodes)
+
+        if preferred_nodes is not None:
+            self.preferred_nodes = preferred_nodes
+        else:
+            self.logger.debug('Operating with no preferred nodes')
```
Comment (pipermerriam) [inline p2p/discovery.py]: Might be worth going ahead and doing a debug log in the other conditional branch to log what preferred nodes we are using.

File: p2p/discovery.py [hunk 7]
```diff
@@ -234,9 +247,134 @@ def send_neighbours(self, node: kademlia.Node, neighbours: List[kademlia.Node])
...
+    async def _run(self) -> None:
+        connect_loop_sleep = 2
+        asyncio.ensure_future(self.proto.bootstrap())
+        while True:
+            await self.maybe_connect_to_more_peers()
+            await self.wait(asyncio.sleep(connect_loop_sleep))
+
+    async def maybe_connect_to_more_peers(self) -> None:
+        """Connect to more peers if we're not yet maxed out to max_peers"""
+        num_connected_nodes = len(self.peer_pool)
```
Comment (pipermerriam) [inline p2p/discovery.py]: Can you move this inside the `if` statement (or maybe just inline with the logging statement)

File: p2p/discovery.py (hunk 6 above)
Comment (gsalgado) [reply p2p/discovery.py]: I'll move it outside of the if block, always logging self.preferred_nodes, and change it into a INFO as this is only going to show up once at startup

**Pull request merged**